### PR TITLE
fix: sequence read rate is slow when use filer.ChunkReadAt.ReadAt

### PR DESCRIPTION
**Pull request by ningfdx**
### When I use mount fuse to test the io rate, I got a wired result. The reading sequence is very slow。#3069 
Screenshot from 2022-05-17 16-14-11

### So i analyze some pprof of the weed mount program. Got this: 
Screenshot from 2022-05-18 17-59-50
Screenshot from 2022-05-18 18-00-19

### Pull Request Description
  The filer.(*ChunkReadAt).ReadAt cost too much time, because every read request call the filer.fetchChunkRange. 

  I read the code of filer.(*ChunkReadAt).ReadAt, and found the reason is filer.ChunkReadAt.readerPattern.isStreaming was set to false, and can not be true forever. 

  That make the every fs.Read() make a ChunkRange request to volume server, so the read rate is very slow.

  Compare the read speed after this pull request:
Screenshot from 2022-05-18 19-11-15

  And I use fio test the random read 4K, the performance is same as origin version weed.

### CODE REVIEW

File: weed/filer/filechunk_manifest.go [hunk 1]
```diff
@@ -115,7 +116,12 @@ func fetchWholeChunk(bytesBuffer *bytes.Buffer, lookupFileIdFn wdclient.LookupFi
 	return nil
 }
 
+var ErrNilFuncNotPermitted = errors.New("nil func is not permitted")
+
 func fetchChunkRange(buffer []byte, lookupFileIdFn wdclient.LookupFileIdFunctionType, fileId string, cipherKey []byte, isGzipped bool, offset int64) (int, error) {
+	if lookupFileIdFn == nil {
```
Comment (chrislusf) [inline weed/filer/filechunk_manifest.go:122]: this is not related?

File: weed/filer/reader_at.go [hunk 2]
```diff
@@ -98,10 +98,9 @@ func (c *ChunkReadAt) Close() error {
 
 func (c *ChunkReadAt) ReadAt(p []byte, offset int64) (n int, err error) {
 
-	c.readerPattern.MonitorReadAt(offset, len(p))
-
 	c.readerLock.Lock()
 	defer c.readerLock.Unlock()
+	c.readerPattern.MonitorReadAt(offset, len(p))
```
Comment (chrislusf) [inline weed/filer/reader_at.go:103]: why this is needed?

File: weed/filer/reader_at_test.go [hunk 3]
```diff
@@ -163,3 +163,86 @@ func TestReaderAt1(t *testing.T) {
...
+	}
+
+	testReadAt(t, readerAt, 0, 10, 10, io.EOF)
+	testReadAt(t, readerAt, 1, 45, 9, io.EOF)
+	testReadAt(t, readerAt, 42, 8, 0, io.EOF)
+	testReadAt(t, readerAt, 3, 6, 6, nil)
+
+	// we don't implement rand read api in test, if c.readerPattern.IsRandomMode() == true, test should get this error
+	readerAt.readerPattern = NewReaderPattern()
+	testReadAtWithMonitorReadAt(t, readerAt, 2, 8, 0, ErrNilFuncNotPermitted)
```
Comment (chrislusf) [inline weed/filer/reader_at_test.go:235]: the tests only check errors are valid. How the data are checked?

Comment (ningfdx) [reply weed/filer/reader_at_test.go:235]: Well, This pr aims at the judgement of the RandomRead in SeaweedFS.Mount client.

I read the file read flow of SeaweedFS.Mount, The single read range [start, end] flow is:
```
1. if the read op is a random read, just fetch VolumeServer over HTTP request with Header 'Range: bytes=start-end', return
2. if the read op is not a random read, caching the hole chunk from VolumeServer, return [start, end] content.
```

#### To check the data, I should write a unit test to verify the range interface and file cache interface of the volume server, I think seaweedfs/weed/filer is not the suitable place to add the unit test.

So I write the unit test to cover the case of "Corrective Judgment", and this unit test is not aims at detect the error of ErrNilFuncNotPermitted.
The error ErrNilFuncNotPermitted will never happen beside unit test. 

This error is only use in unit test to detect the flow of code branch execution
```
1. func testReadAtWithMonitorReadAt() will return ErrNilFuncNotPermitted if a doReadAt() is a random read.
2. func testReadAtWithMonitorReadAt() will return nil/io.EOF if a doReadAt() is a none random read.
```
Please let me know if my understanding is wrong :)

File: weed/filer/reader_at.go (hunk 2 above)
Comment (ningfdx) [reply weed/filer/reader_at.go:103]: the MonitorReadAt is not concurrent safe, I think call this func after lock is high Robustness.
By the way, I don't know whether fuse will call multi read at same time, could this happen？

**Comment by chrislusf** [comment]
I used your code and added ebd6097e6f7c60599e198e5444cdcb353b6328b7
